src/Symfony/Component/Intl/Resources/emoji/build.php

```diff
@@ -0,0 +1,135 @@
+use Symfony\Component\Finder\Finder;
+
+Builder::cleanTarget();
+$emojisCodePoints = Builder::getEmojisCodePoints();
+Builder::saveRules(Builder::buildRules($emojisCodePoints));
+
+final class Builder
+{
+    private const TARGET_DIR = __DIR__.'/../data/transliterator/emoji/';
+
+    public static function getEmojisCodePoints(): array
+    {
+        $lines = file(__DIR__.'/vendor/unicode-org/cldr/tools/cldr-code/src/main/resources/org/unicode/cldr/util/data/emoji/emoji-test.txt');
+
+        $emojisCodePoints = [];
+        foreach ($lines as $line) {
+            $line = trim($line);
+            if (!$line || str_starts_with($line, '#')) {
+                continue;
+            }
+
+            // 263A FE0F    ; fully-qualified     # ☺️ E0.6 smiling face
+            preg_match('{^(?<codePoints>[\w ]+) +; [\w-]+ +# (?<emoji>.+) E\d+\.\d+ ?(?<name>.+)$}Uu', $line, $matches);
+            if (!$matches) {
+                throw new \DomainException("Could not parse line: \"$line\".");
+            }
+
+            $codePoints = strtolower(trim($matches['codePoints']));
+            $emojisCodePoints[$codePoints] = $matches['emoji'];
+            // We also add a version without the "Zero Width Joiner"
+            $codePoints = str_replace('200d ', '', $codePoints);
+            $emojisCodePoints[$codePoints] = $matches['emoji'];
+        }
+
+        return $emojisCodePoints;
+    }
+
+    public static function buildRules(array $emojisCodePoints): Generator
+    {
+        $files = (new Finder())
+            ->files()
+            ->in([
+                __DIR__.'/vendor/unicode-org/cldr/common/annotationsDerived',
+                __DIR__.'/vendor/unicode-org/cldr/common/annotations',
+            ])
+            ->name('*.xml')
+        ;
+
+        $ignored = [];
+        $mapsByLocale = [];
+
+        foreach ($files as $file) {
+            $locale = $file->getBasename('.xml');
+
+            $document = new DOMDocument();
+            $document->loadXML(file_get_contents($file));
+            $xpath = new DOMXPath($document);
+            $results = $xpath->query('.//annotation[@type="tts"]');
+
+            foreach ($results as $result) {
+                $emoji = $result->getAttribute('cp');
+                $name = $result->textContent;
+                $parts = preg_split('//u', $emoji, -1, \PREG_SPLIT_NO_EMPTY);
+                $emojiCodePoints = implode(' ', array_map('dechex', array_map('mb_ord', $parts)));
+                if (!array_key_exists($emojiCodePoints, $emojisCodePoints)) {
+                    $ignored[] = [
+                        'locale' => $locale,
+                        'emoji' => $emoji,
+                        'name' => $name,
+                    ];
+                    continue;
+                }
+                self::testEmoji($emoji, $locale);
+                $codePointsCount = mb_strlen($emoji);
+                $mapsByLocale[$locale][$codePointsCount][$emoji] = $name;
+            }
```

Comment:
Why not yielding after the nested `foreach` with a scoped `$maps` variable instead of keeping everything in memory through `$mapsByLocale`?

Reply:
Because we need to compute the whole $locale space in one shot.
There are two files to parse before being able to "yield" it :)

Reply:
I see, thanks!